### PR TITLE
Entity type: loader caching by query builder instance

### PR DESCRIPTION
**Pull request by dominikzogg**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets |  |
| License | MIT |
| Doc PR |  |

Using entity types within a collection, can get very slow, if a query_builder option is set.

Without this option, the doctrine type handles everything itself, and makes it calls against the manager registry, but using this, calls the getLoader method, which ends in build the choice list again for each collection element, which makes more than twice the time for a page load within my energycalculator.

https://github.com/dominikzogg/symfony/blob/2.7/src/Symfony/Bridge/Doctrine/Form/Type/DoctrineType.php#L68

https://github.com/dominikzogg/symfony/blob/2.7/src/Symfony/Bridge/Doctrine/Form/Type/DoctrineType.php#L108

https://github.com/dominikzogg/energycalculator/blob/master/src/Dominikzogg/EnergyCalculator/Form/ComestibleWithinDayType.php

### CODE REVIEW

**Comment by dominikzogg** [comment]
@Tobion, i made a solution which caches only if the query builder is a query builder instance

File: src/Symfony/Bridge/Doctrine/Form/Type/EntityType.php [hunk 1]
```diff
@@ -27,11 +33,60 @@ class EntityType extends DoctrineType
...
+    }
+
+    /**
+     * @param QueryBuilder $queryBuilder
+     *
+     * @return string
+     */
+    protected function getQueryBuilderHash(QueryBuilder $queryBuilder)
+    {
+        return hash('sha256', json_encode(array(
```
Comment (timglabisch) [inline src/Symfony/Bridge/Doctrine/Form/Type/EntityType.php]: couldn't we provide something faster than sha256? is this really needed here?

File: src/Symfony/Bridge/Doctrine/Form/Type/EntityType.php [hunk 2]
```diff
@@ -27,11 +33,60 @@ class EntityType extends DoctrineType
...
-            $manager,
-            $class
-        );
+        $queryBuilderHash = null;
+
+        if ($queryBuilder instanceof QueryBuilder) {
+            $queryBuilderHash = $this->getQueryBuilderHash($queryBuilder);
+        }
+
+        if (null === $queryBuilderHash) {
```
Comment (timglabisch) [inline src/Symfony/Bridge/Doctrine/Form/Type/EntityType.php]: the `queryBuilderhash` must not be null if `$queryBuilder instanceof QueryBuilder` i dont like the structure of the if conditions here.

File: src/Symfony/Bridge/Doctrine/Form/Type/EntityType.php (hunk 1 above)
Comment (dominikzogg) [inline src/Symfony/Bridge/Doctrine/Form/Type/EntityType.php]: @timglabisch well i don't like it, but i made it as in DoctrineType: https://github.com/symfony/symfony/blob/2.7/src/Symfony/Bridge/Doctrine/Form/Type/DoctrineType.php#L117

Comment (timglabisch) [inline src/Symfony/Bridge/Doctrine/Form/Type/EntityType.php]: ah great :)

**Comment by dominikzogg** [comment]
@timglabisch any idea, why the php5.3 test against low deps fails? The output on travis doesn't help

https://travis-ci.org/symfony/symfony/jobs/53549996

**Comment by xabbuh** [comment]
@dominikzogg This was just one of the occasionally failing test cases and is not related to your changes.

**Comment by dominikzogg** [comment]
@xabbuh ah ok, thank you

**Comment by dominikzogg** [comment]
@stof is it possible to milestone this for 2.7?

File: src/Symfony/Bridge/Doctrine/Form/Type/EntityType.php [hunk 3]
```diff
@@ -27,11 +33,55 @@ class EntityType extends DoctrineType
...
+
+        return $this->loaderCache[$loaderHash];
+    }
+
+    /**
+     * @param QueryBuilder $queryBuilder
+     *
+     * @return string
+     */
+    protected function getQueryBuilderHash(QueryBuilder $queryBuilder)
```
Comment (fabpot) [inline src/Symfony/Bridge/Doctrine/Form/Type/EntityType.php]: This one should be private.

File: src/Symfony/Bridge/Doctrine/Form/Type/EntityType.php [hunk 4]
```diff
@@ -27,11 +33,55 @@ class EntityType extends DoctrineType
...
+    }
+
+    /**
+     * @param ObjectManager $manager
+     * @param string        $queryBuilderHash
+     * @param string        $class
+     *
+     * @return string
+     */
+    protected function getLoaderHash(ObjectManager $manager, $queryBuilderHash, $class)
```
Comment (fabpot) [inline src/Symfony/Bridge/Doctrine/Form/Type/EntityType.php]: should be private as well.

**Comment by dominikzogg** [comment]
@fabpot done

**Comment by fabpot** [comment]
Thank you @dominikzogg.

**Comment by dominikzogg** [comment]
@fabpot thank you

File: src/Symfony/Bridge/Doctrine/Form/Type/EntityType.php [hunk 5]
```diff
@@ -27,11 +33,55 @@ class EntityType extends DoctrineType
...
+    /**
+     * @param QueryBuilder $queryBuilder
+     *
+     * @return string
+     */
+    private function getQueryBuilderHash(QueryBuilder $queryBuilder)
+    {
+        return hash('sha256', json_encode(array(
+            'sql' => $queryBuilder->getQuery()->getSQL(),
+            'parameters' => $queryBuilder->getParameters(),
```
Comment (stof) [inline src/Symfony/Bridge/Doctrine/Form/Type/EntityType.php]: parameters can contain objects (entities or DateTime objects for instance), which will not work well here

**Comment by stof** [comment]
Most of the type, the ``query_builder`` option is used with a closure returning the QueryBuilder (and receiving the entity repository as argument) rather than by passing a QueryBuilder directly (allowing to avoid injecting the Doctrine layer in your parent form type). This means that your cache will not be applied in most cases

**Comment by dominikzogg** [comment]
@stof well i tried to get a solution to compare those closures, but its much easier to get this stable.

File: src/Symfony/Bridge/Doctrine/Form/Type/EntityType.php (hunk 5 above)
Comment (dominikzogg) [inline src/Symfony/Bridge/Doctrine/Form/Type/EntityType.php]: @stof \DateTime would be possible yeah, but other objects aren't, cause on relation the id and not the object is needed, or did this change?

Comment (dominikzogg) [inline src/Symfony/Bridge/Doctrine/Form/Type/EntityType.php]: @stof it looks like that the only object is allowed as a parameter is a datetime object, which is easy to json_encode, which means this PR should work fine https://github.com/doctrine/doctrine2/blob/master/lib/Doctrine/ORM/Query/ParameterTypeInferer.php#L66

**Pull request merged**